pallets/crowdloan-reward/src/lib.rs

```diff
@@ -607,13 +580,26 @@ impl<T: Config> Reward for Pallet<T>
 
     let to = <T::Lookup as StaticLookup>::unlookup(who.clone());
 
+
+    T::Currency::transfer(&from, &who, direct_reward, KeepAlive)?;
+
+    // Currently I know no way to secure that both extrinsic (transfer, vested_transfer)
+    // will be successful or be reverted if one of them changes.
+    // So, as `vested_transfer` is not revertible, we first transfer the direct amount, and then
+    // the vested amount. If first fails, we simply abort. If second fails, we are transferring
+    // the direct payout back to the module.
+    //
+    // NOTE: This procedure does change the state...
     <pallet_vesting::Module<T>>::vested_transfer(
       T::Origin::from(RawOrigin::Signed(from.clone())),
       to,
       schedule
-    )?;
+    ).map_err(|err| {
+      T::Currency::transfer(&who, &from, direct_reward, AllowDeath)
```

Comment:
`KepAlive` instead

Reply:
I think we need to allow death here. The account of the contributor is possibly a fresh one and if we do not allow death, we always loose a small amount upon errors of the vested transfer call.

Reply:
true but this transfer doesn't involve any fees so we pretty negated the action. So if there are any dust, it pretty belongs to the `who`, no ?

Reply:
Yes, of course. But in the case of a newly created account, the rollback of the previous transfer will fail of we use `KeepAlive` due to this [code line](https://github.com/paritytech/substrate/blob/75957575080a6eacfbdf8995b63f2693ee33efe6/frame/balances/src/lib.rs#L1351), leaving the direct payout at `who`.  As the claim module will register this as a failed claim, it will allow the contributor to claim again. If he does so with another newly created account he could get the direct payout a second time, and so forth.

Basically what I mean here is: If the account is above the existential deposit solely due to our previous transfer, we must be able to purge him, otherwise, we don't get our funds back.

Another thing that could be problematic is this [code line](https://github.com/paritytech/substrate/blob/75957575080a6eacfbdf8995b63f2693ee33efe6/frame/balances/src/lib.rs#L1350):
`!system::Pallet::<T>::is_provider_required(transactor)`, which checks if the account has any other module being dependent on it.

Do you know if there is the possibility for an account to have such a `consumers` reference while having less than the existential deposit as a balance? Then again, the scenario from above would apply.

Reply:
I updated the code. It now checks in advance, if a vesting schedule exists. Currently, this and the amount to low error, are the only reason a vested transfer might fail. As we already check upfront if the module has enough funds for this reward payout. 

Generally, I would say that there is no reason that one of the transfers should fail now. We could therefore think about removing the action upon failure of the `vested_transfer`. But maybe better be safe than sorry? Of course, we then still have to decide, the above case.

Reply:
Since the `keepAlive|AllowDeath` are only applicable to the from, then is up to us to always have enough funds to cover the existential deposit, and allowing death for a module_id account is prob not a good idea.
Maybe we can add the `existential_deposit` to the total amount available check above, so:
```
ensure!( free_balance > reward + T::ExistentialDeposit::get(), Error::<T>::NotEnoughFunds );
```
Then we cover the case of keeping alive, right? 

For the transactor is_provider check then we do not need to care about it since the account will stay alive.

Reply:
> Since the `keepAlive|AllowDeath` are only applicable to the from, then is up to us to always have enough funds to cover the existential deposit, and allowing death for a module_id account is prob not a good idea.
Why is it only applicable to the `from`? In case, we want to revert the previous transfer, it is applied to the `who` (recipient account).

> Maybe we can add the `existential_deposit` to the total amount available check above, so:
This check is done at line 560. The module also transfers with `KeepAlive` to the recipient. So, we should be good on this side.

> For the transactor is_provider check then we do not need to care about it since the account will stay alive.
But this will also mean, that the transfer fails and we can not revert the previous transfer. Although, I think we should be safe, as we are checking everything upfront and now.